### ISSUE TITLE
Soundness issue in association establishment when association acceptor sends more PDUs in quick succession

### ISSUE THREAD

**Issue by Enet4** [Enet4/dicom-rs#589]
Stub, to be expanded later

**Comment by naterichman** [comment]
Getting back here after a while, sorry for introducing this regression!  What is left to be done here, can I help?

**Comment by Enet4** [comment]
This is [still an issue](https://github.com/Enet4/dicom-rs/blob/master/ul/src/association/client.rs#L690), though I'm not sure how much of an impact this has. So far I haven't stumbled upon any failures in association establishment.

I am honestly a bit more worried about `get_client_pdu_async`, which needs its temporary byte buffer plucked out onto the caller otherwise it can lose data between calls. This is something that should probably earn more attention, as it could put most async code at risk. Please let me know if you are available and willing to take either issue @naterichman . :)

**Comment by naterichman** [comment]
I will try to take a look at both at some point! I'm pretty busy right now so it might not be for a month or so, but I think I'll first try to make a test to consistently reproduce that for both sync and async code.

**Comment by naterichman** [comment]
Sorry for the delay, things have been hectic, but I'm finally getting back to looking at this.  Based on my reading it seems like this could happen for both sync and async code and both on the SCU or SCP side during establishment (on the SCU side if the SCP sends AssociationAC and one or more other PDUs, and on the SCP side if the SCU sends the AssociationRQ and one or more other PDUs), since a temporary buffer is created in both the Client::establish and Server::establish and then dropped, is that right?

I'm trying to understand what to do about it, it seems like we could just add another buffer onto both the `ServerAssociation` and `ClientAssociation` structs and pass any remaining data into that, but then what should the `Server/Client::Association` do with that? On the SCU side, what would be the expected behavior for the SCP to send the Acknowledge and then one or more PDUs, it should be the SCU that is initiating requests to the SCP.  And on the SCP side, it seems like user error to send both the request and other PDUs before receiving an acknowledge.

I'm, not sure if I'm making sense, or even fully understanding the potential problem here.

**Comment by Enet4** [comment]
> Sorry for the delay, things have been hectic, but I'm finally getting back to looking at this. Based on my reading it seems like this could happen for both sync and async code and both on the SCU or SCP side during establishment (on the SCU side if the SCP sends AssociationAC and one or more other PDUs, and on the SCP side if the SCU sends the AssociationRQ and one or more other PDUs), since a temporary buffer is created in both the Client::establish and Server::establish and then dropped, is that right?

Yup, so the warning is in the synchronous API, but this problem also applies to the async API in a way which requires changing `get_client_pdu_async` (to receive a buffer so it can be reused and no data is lost).

> I'm trying to understand what to do about it, it seems like we could just add another buffer onto both the `ServerAssociation` and `ClientAssociation` structs and pass any remaining data into that, but then what should the `Server/Client::Association` do with that? On the SCU side, what would be the expected behavior for the SCP to send the Acknowledge and then one or more PDUs, it should be the SCU that is initiating requests to the SCP. And on the SCP side, it seems like user error to send both the request and other PDUs before receiving an acknowledge.

I had another look at the code, and for the sync API, most of heavy lifting is already done.  `ClientAssociation` and `ServerAssociation` already have a reading buffer. It might be enough to keep using the buffer created on `establish_impl`, which means passing `buf` to the new `ClientAssociation` returned (client.rs:756). Then we need to test that everything continues to work as expected.

For the async API, we first need to change `get_client_pud_async`, then do the equivalent in `establish_impl_async` (client.rs:1313).
